src/libcore/borrow.rs

```diff
@@ -12,26 +12,157 @@
 
 #![stable(feature = "rust1", since = "1.0.0")]
 
-/// A trait for borrowing data.
+/// A trait identifying how borrowed data behaves.
 ///
-/// In general, there may be several ways to "borrow" a piece of data.  The
-/// typical ways of borrowing a type `T` are `&T` (a shared borrow) and `&mut T`
-/// (a mutable borrow). But types like `Vec<T>` provide additional kinds of
-/// borrows: the borrowed slices `&[T]` and `&mut [T]`.
+/// In Rust, it is common to provide different representations of a type for
+/// different use cases. For instance, storage location and management for a
+/// value can be specifically chosen as appropriate for a particular use via
+/// pointer types such as [`Box<T>`] or [`Rc<T>`] or one can opt into
+/// concurrency via synchronization types such as [`Mutex<T>`], avoiding the
+/// associated cost when in parallel doesn’t happen. Beyond these generic
+/// wrappers that can be used with any type, some types provide optional
+/// facets providing potentially costly functionality. An example for such a
+/// type is [`String`] which adds the ability to extend a string to the basic
+/// [`str`]. This requires keeping additional information unnecessary for a
+/// simple, imutable string.
 ///
-/// When writing generic code, it is often desirable to abstract over all ways
-/// of borrowing data from a given type. That is the role of the `Borrow`
-/// trait: if `T: Borrow<U>`, then `&U` can be borrowed from `&T`.  A given
-/// type can be borrowed as multiple different types. In particular, `Vec<T>:
-/// Borrow<Vec<T>>` and `Vec<T>: Borrow<[T]>`.
+/// These types signal that they are a specialized representation of a basic
+/// type `T` by implementing `Borrow<T>`. The method `borrow` provides a way
+/// to convert a reference to the type into a reference to the underlying
+/// basic type.
 ///
-/// If you are implementing `Borrow` and both `Self` and `Borrowed` implement
-/// `Hash`, `Eq`, and/or `Ord`, they must produce the same result.
+/// If a type implementing `Borrow<T>` implements other traits also
+/// implemented by `T`, these implementations behave identically if the trait
+/// is concerned with the data rather than its representation. For instance,
+/// the comparison traits such as `PartialEq` or `PartialOrd` must behave
+/// identical for `T` and any type implemeting `Borrow<T>`.
 ///
-/// `Borrow` is very similar to, but different than, `AsRef`. See
-/// [the book][book] for more.
+/// When writing generic code, a use of `Borrow` should always be justified
```

Comment:
Alternative proposal: ‘When using `Borrow` as a trait bound’? And tone down the text to be more along the line of suggesting instead of demanding?

Reply:
That seems reasonable. It may be appropriate to phrase more as justification than as instruction. Instead of instructing a particular use: "here is how you should always do it" or "here is how we suggest that you do it," more like a justification "here is how you will typically find it used and why."